arraycontext/container/dataclass.py

```diff
@@ -104,33 +114,46 @@ def is_array_field(f: Field) -> bool:
                 _BaseGenericAlias,
                 _SpecialForm,
             )
-            if isinstance(f.type, GenericAlias | _BaseGenericAlias | _SpecialForm):
+            if isinstance(field_type, GenericAlias | _BaseGenericAlias | _SpecialForm):
                 # NOTE: anything except a Union is not allowed
                 raise TypeError(
                         f"Typing annotation not supported on field '{f.name}': "
-                        f"'{f.type!r}'")
+                        f"'{field_type!r}'")
 
-            if not isinstance(f.type, type):
+            if not isinstance(field_type, type):
                 raise TypeError(
                         f"Field '{f.name}' not an instance of 'type': "
-                        f"'{f.type!r}'")
-
-        return is_array_type(f.type)
-
-    from pytools import partition
-    array_fields, non_array_fields = partition(is_array_field, fields(cls))
+                        f"'{field_type!r}'")
+
+        return is_array_type(field_type)
+
+    array_fields: list[Field] = []
+    non_array_fields: list[Field] = []
+    from inspect import get_annotations
+    cls_ann: Mapping[str, type] | None = None
+    for field in fields(cls):
+        field_type_or_str = field.type
+        if isinstance(field_type_or_str, str):
+            if cls_ann is None:
+                cls_ann = get_annotations(cls, eval_str=True)
+            field_type = cls_ann[field.name]
+        else:
+            field_type = field_type_or_str
+        (array_fields
+            if is_array_field(field, field_type)
+            else non_array_fields).append(field)
```

Comment:
This is a bit ugly.. maybe a normal if+else?

Reply:
Fixed!